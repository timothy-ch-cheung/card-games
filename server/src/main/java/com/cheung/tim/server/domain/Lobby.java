package com.cheung.tim.server.domain;

import com.cheung.tim.server.enums.GameMode;
import com.cheung.tim.server.enums.GameStatus;
import lombok.Getter;
import lombok.Setter;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.Collections;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

@Entity
@Table(name = "LOBBY")
public class Lobby extends BaseEntity {

    public Lobby() {
    }

    public Lobby(String lobbyName, Player host, GameStatus status, Integer maxPlayers, GameMode gameMode) {
        this.lobbyName = lobbyName;
        this.host = host;
        this.gameStatus = status;
        this.maxPlayers = maxPlayers;
        this.gameMode = gameMode;
        this.rounds = gameMode.getInitialRounds();
    }


    @Id
    @Getter
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long lobbyId;

    @Getter
    @Setter
    private String lobbyName;

    @Getter
    @Setter
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "game_id", unique = true)
    private Game game;

    @Getter
    @Setter
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", unique = true)
    private Player host;

    @OneToMany(
            mappedBy = "currentLobby",
            cascade = CascadeType.PERSIST,
            fetch = FetchType.LAZY)
    private Set<Player> guests = new HashSet<>();

    @Getter
    @Setter
    @Enumerated(EnumType.STRING)
    private GameStatus gameStatus;

    @Getter
    @Setter
    @Enumerated(EnumType.STRING)
    private GameMode gameMode;

    @Getter
    @Setter
    @NotNull
    private Integer maxPlayers;

    @Getter
    @Setter
    private Integer rounds;

    public Set<Player> getGuests() {
        return Collections.unmodifiableSet(this.guests);
    }

    public void addGuest(Player player) {
        this.guests.add(player);
    }

    @Override
    public int hashCode() {
        return Objects.hash(lobbyId, lobbyName, host, guests, gameStatus);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }
        Lobby otherLobby = (Lobby) obj;
        return Objects.equals(lobbyId, otherLobby.lobbyId) &&
                Objects.equals(lobbyName, otherLobby.lobbyName) &&
                Objects.equals(host, otherLobby.host) &&
                Objects.equals(guests, otherLobby.guests) &&
                Objects.equals(gameStatus, otherLobby.gameStatus);
    }
}
