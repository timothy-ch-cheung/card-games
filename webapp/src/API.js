import axios from 'axios';

const instance = axios.create({
    baseURL: 'http://127.0.0.1:8080'
});

export default instance;