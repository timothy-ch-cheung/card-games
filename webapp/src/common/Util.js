import {act} from "@testing-library/react";

export function isBlank(str) {
    return (!str || /^\s*$/.test(str));
}

export const flushPromises = () => new Promise(resolve => setTimeout(resolve))

function wait(amount = 0) {
    return new Promise(resolve => setTimeout(resolve, amount));
}

export async function actWait(amount = 0) {
    await act(async () => {
        await wait(amount);
    });
}