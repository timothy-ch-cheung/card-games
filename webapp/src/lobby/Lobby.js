import React, {useEffect, useState} from "react";
import {useDispatch, useSelector} from "react-redux";
import API from "../API";
import {useHistory} from "react-router-dom";
import {resetGame, resetGameMode, setGameMode} from "../redux/actions";
import PlayerList from "../components/player-list/PlayerList";
import LobbySettings from "../components/lobby-settings/LobbySettings";

function Lobby(props) {
    const gameId = useSelector(state => state.game);
    const userId = useSelector(state => state.user);
    const userKey = useSelector(state => state.key);
    const gameMode = useSelector(state => state.gameMode);
    const [game, setGame] = useState({});
    const history = useHistory();
    const dispatch = useDispatch();

    const getPlayersList = () => {
        let host = game.host || {};
        host.isHost = true;
        return game.guests ? [host].concat(game.guests) : [host];
    }

    useEffect(() => {
        const getGame = () => {
            API.get(`/game/${gameId}`
            ).then(function (response) {
                if (response.data.gameStatus === 'DELETED') {
                    dispatch(resetGame());
                    clearInterval(interval);
                    history.push("/games/public");
                }
                setGame(response.data);
                setGameMode(response.data.gameMode);
            }).catch(function (error) {
                props.onShowError(error.response.data.message);
                clearInterval(interval);
                history.push("/games/public");
            });
        };

        getGame();
        const interval = setInterval(function () {
            getGame();
        }, 1000);

        return () => {
            clearInterval(interval);
        }
    }, [gameId, history, props, dispatch]);

    const onLeaveGame = () => {
        API.patch(`/leave/${gameId}`, {
            id: userId,
            key: userKey
        }).then(function (response) {
            dispatch(resetGame());
            dispatch(resetGameMode());
            history.push('/games/public');
        }).catch(function (error) {
            console.log(error);
        });
    }

    return (
        <div style={{display: "flex"}}>
            <PlayerList players={getPlayersList()} onLeave={onLeaveGame} numPlayers={2} maxPlayers={game.maxPlayers}/>
            <LobbySettings gameMode={gameMode} numPlayers={2}/>
        </div>
    );

}

export default Lobby;