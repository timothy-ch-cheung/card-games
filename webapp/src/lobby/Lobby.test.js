import React from 'react';
import {configure, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Lobby from "./Lobby";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import API from "../API";
import {flushPromises} from "../common/Util";

configure({adapter: new Adapter()});

jest.mock('react-router-dom', () => ({
    useHistory: () => ({
        push: jest.fn(),
    }),
}));

describe("TEST SUITE Lobby: ", () => {
    let wrapper;
    const showError = jest.fn();
    const mockStore = configureStore([]);
    let store = mockStore({gameMode: "MATCH_TWO", game: 1});
    const MockAdapter = require("axios-mock-adapter");
    let mockAPI;

    beforeAll(() => {
        mockAPI = new MockAdapter(API);
    });

    afterEach(() => {
        store.clearActions();
        mockAPI.resetHistory();
        mockAPI.reset();
    });

    describe("When lobby is active, ", () => {
        beforeEach(() => {
            mockAPI.onGet("/game/1").reply(200, {
                id: 1,
                host: {id: "123", username: "John"},
                guests: [{id: "321", username: "Jane"}],
                gameMode: "MATCH_TWO",
                gameStatus: "OPEN"
            });
            wrapper = mount(<Provider store={store}><Lobby onShowError={showError}/></Provider>);
        });

        test('Matches Initial Lobby snapshot', () => {
            expect(wrapper).toMatchSnapshot();
            wrapper.unmount();
        });

        test('Leave game button sends PATCH', async () => {
            mockAPI.onPatch("/leave/1").reply(204, {});
            wrapper.find('button[data-test="leave-game-btn"]').simulate('click');
            await flushPromises();

            expect(mockAPI.history.patch.length).toBe(1);
            let actions = store.getActions();
            expect(actions.length).toBe(3);
            expect(actions[0]).toEqual({"type": "SET_GAME_MODE", "payload": "MATCH_TWO"});
            expect(actions[1]).toEqual({"type": "RESET_GAME"});
            expect(actions[2]).toEqual({"type": "RESET_GAME_MODE"});
            wrapper.unmount();
        });
    });

    describe("When lobby is DELETED", () => {
        beforeEach(() => {
            mockAPI.onGet("/game/1").reply(200, {
                id: 1,
                host: {id: "123", username: "John"},
                guests: [{id: "321", username: "Jane"}],
                gameMode: "MATCH_TWO",
                gameStatus: "DELETED"
            });
            wrapper = mount(<Provider store={store}><Lobby onShowError={showError}/></Provider>);
        });

        test('On component load GET DELETED game resets Game Id', async () => {
            await flushPromises();
            let actions = store.getActions();
            expect(actions.length).toBe(1);
            expect(actions[0]).toEqual({"type": "RESET_GAME"});
        });
    });

    describe("When lobby is not returned ", () => {

        beforeEach(() => {
            mockAPI.onGet("/game/1").reply(400, {
                message: "Server inactive"
            });
            wrapper = mount(<Provider store={store}><Lobby onShowError={showError}/></Provider>);
        });

        test('from API call after component mount', async () => {
            await flushPromises();
            let actions = store.getActions();
            expect(actions.length).toBe(1);
            expect(actions[0]).toEqual({"type": "RESET_GAME"});
        });
    });
});