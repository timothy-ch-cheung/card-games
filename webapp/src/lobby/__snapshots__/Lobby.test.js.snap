// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST SUITE Lobby:  When lobby is active,  Matches Initial Lobby snapshot 1`] = `
<Provider
  onShowError={[MockFunction]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Lobby>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <SockJsClient
        autoReconnect={true}
        debug={false}
        getRetryInterval={[Function]}
        headers={Object {}}
        heartbeat={10000}
        onConnect={[Function]}
        onDisconnect={[Function]}
        onMessage={[Function]}
        options={Object {}}
        subscribeHeaders={Object {}}
        topics={
          Array [
            "/topic/game/1",
          ]
        }
        url="https://card-games-tcc.herokuapp.com/websocket"
      />
      <PlayerList
        onLeave={[Function]}
        players={
          Array [
            Object {
              "isHost": true,
            },
          ]
        }
      >
        <Card
          body={false}
          style={
            Object {
              "margin": "10px",
              "width": "35%",
            }
          }
        >
          <div
            className="card"
            style={
              Object {
                "margin": "10px",
                "width": "35%",
              }
            }
          >
            <CardHeader
              style={
                Object {
                  "display": "flex",
                  "justifyContent": "space-between",
                  "paddingLeft": "10px",
                }
              }
            >
              <div
                className="card-header"
                style={
                  Object {
                    "display": "flex",
                    "justifyContent": "space-between",
                    "paddingLeft": "10px",
                  }
                }
              >
                <p
                  style={
                    Object {
                      "marginBottom": "0",
                    }
                  }
                >
                  Lobby
                </p>
                <p
                  data-test="player-count"
                  style={
                    Object {
                      "marginBottom": "0",
                    }
                  }
                >
                  1
                  /
                </p>
              </div>
            </CardHeader>
            <CardBody
              style={
                Object {
                  "height": "400px",
                  "overflowY": "auto",
                  "padding": "5px 5px",
                }
              }
            >
              <div
                className="card-body"
                style={
                  Object {
                    "height": "400px",
                    "overflowY": "auto",
                    "padding": "5px 5px",
                  }
                }
              >
                <ForwardRef
                  borderless={true}
                  variant="light"
                >
                  <table
                    className="table table-light table-borderless"
                  >
                    <tbody />
                  </table>
                </ForwardRef>
              </div>
            </CardBody>
            <CardBody
              style={
                Object {
                  "borderTop": "1px solid #dfdfdf",
                  "padding": "10px",
                }
              }
            >
              <div
                className="card-body"
                style={
                  Object {
                    "borderTop": "1px solid #dfdfdf",
                    "padding": "10px",
                  }
                }
              >
                <Button
                  active={false}
                  className="leave-game"
                  data-test="leave-game-btn"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                  variant="info"
                >
                  <button
                    className="leave-game btn btn-info"
                    data-test="leave-game-btn"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    Leave
                  </button>
                </Button>
              </div>
            </CardBody>
          </div>
        </Card>
      </PlayerList>
      <LobbySettings
        gameId={1}
        gameMode="MATCH_TWO"
        numPlayers={1}
        rounds={1}
      >
        <Card
          body={false}
          style={
            Object {
              "margin": "10px",
              "width": "35%",
            }
          }
        >
          <div
            className="card"
            style={
              Object {
                "margin": "10px",
                "width": "35%",
              }
            }
          >
            <CardHeader
              style={
                Object {
                  "paddingLeft": "10px",
                }
              }
            >
              <div
                className="card-header"
                style={
                  Object {
                    "paddingLeft": "10px",
                  }
                }
              >
                Game: 
                Match Two
              </div>
            </CardHeader>
            <CardBody
              style={
                Object {
                  "height": "400px",
                  "overflowY": "auto",
                  "padding": "5px 10px",
                }
              }
            >
              <div
                className="card-body"
                style={
                  Object {
                    "height": "400px",
                    "overflowY": "auto",
                    "padding": "5px 10px",
                  }
                }
              >
                <Form
                  data-test="lobby-settings-form"
                  inline={false}
                  noValidate={true}
                  onSubmit={[Function]}
                  validated={false}
                >
                  <form
                    className=""
                    data-test="lobby-settings-form"
                    noValidate={true}
                    onSubmit={[Function]}
                  >
                    <RoundPicker>
                      <div
                        style={
                          Object {
                            "display": "flex",
                          }
                        }
                      >
                        <div
                          style={
                            Object {
                              "marginRight": "5px",
                            }
                          }
                        >
                          <FormLabel
                            column={false}
                            srOnly={false}
                          >
                            <label
                              className="form-label"
                            >
                              Stages
                            </label>
                          </FormLabel>
                          <NumberPicker
                            data-test="round-number-picker"
                            disabled={true}
                            name="numRounds"
                            onDecrease={[Function]}
                            onIncrease={[Function]}
                            value={1}
                          >
                            <InputGroup
                              style={
                                Object {
                                  "width": "150px",
                                }
                              }
                            >
                              <div
                                className="input-group"
                                style={
                                  Object {
                                    "width": "150px",
                                  }
                                }
                              >
                                <InputGroupPrepend>
                                  <div
                                    className="input-group-prepend"
                                  >
                                    <Button
                                      active={false}
                                      data-test="numRounds-minus-btn"
                                      disabled={true}
                                      onClick={[Function]}
                                      style={
                                        Object {
                                          "fontSize": "1.5rem",
                                          "padding": "0rem 0.75rem",
                                          "width": "40px",
                                        }
                                      }
                                      type="button"
                                      variant="info"
                                    >
                                      <button
                                        className="btn btn-info"
                                        data-test="numRounds-minus-btn"
                                        disabled={true}
                                        onClick={[Function]}
                                        style={
                                          Object {
                                            "fontSize": "1.5rem",
                                            "padding": "0rem 0.75rem",
                                            "width": "40px",
                                          }
                                        }
                                        type="button"
                                      >
                                        -
                                      </button>
                                    </Button>
                                  </div>
                                </InputGroupPrepend>
                                <FormControl
                                  className="number-picker"
                                  data-test="numRounds-value"
                                  disabled={true}
                                  name="numRounds"
                                  style={
                                    Object {
                                      "backgroundColor": "#fff",
                                      "textAlign": "center",
                                    }
                                  }
                                  value={1}
                                >
                                  <input
                                    className="number-picker form-control"
                                    data-test="numRounds-value"
                                    disabled={true}
                                    name="numRounds"
                                    style={
                                      Object {
                                        "backgroundColor": "#fff",
                                        "textAlign": "center",
                                      }
                                    }
                                    value={1}
                                  />
                                </FormControl>
                                <InputGroupAppend>
                                  <div
                                    className="input-group-append"
                                  >
                                    <Button
                                      active={false}
                                      data-test="numRounds-plus-btn"
                                      disabled={true}
                                      onClick={[Function]}
                                      style={
                                        Object {
                                          "fontSize": "1.5rem",
                                          "padding": "0rem 0.75rem",
                                          "width": "40px",
                                        }
                                      }
                                      type="button"
                                      variant="info"
                                    >
                                      <button
                                        className="btn btn-info"
                                        data-test="numRounds-plus-btn"
                                        disabled={true}
                                        onClick={[Function]}
                                        style={
                                          Object {
                                            "fontSize": "1.5rem",
                                            "padding": "0rem 0.75rem",
                                            "width": "40px",
                                          }
                                        }
                                        type="button"
                                      >
                                        +
                                      </button>
                                    </Button>
                                  </div>
                                </InputGroupAppend>
                              </div>
                            </InputGroup>
                          </NumberPicker>
                        </div>
                        <div
                          style={
                            Object {
                              "display": "visible",
                            }
                          }
                        >
                          <FormLabel
                            column={false}
                            srOnly={false}
                          >
                            <label
                              className="form-label"
                            >
                              Rounds
                            </label>
                          </FormLabel>
                          <FormControl
                            data-test="round-number-text"
                            disabled={true}
                            name="totalRounds"
                            required={true}
                            style={
                              Object {
                                "backgroundColor": "#fff",
                                "textAlign": "center",
                                "width": "80px",
                              }
                            }
                            type="text"
                            value={2}
                          >
                            <input
                              className="form-control"
                              data-test="round-number-text"
                              disabled={true}
                              name="totalRounds"
                              required={true}
                              style={
                                Object {
                                  "backgroundColor": "#fff",
                                  "textAlign": "center",
                                  "width": "80px",
                                }
                              }
                              type="text"
                              value={2}
                            />
                          </FormControl>
                        </div>
                      </div>
                    </RoundPicker>
                  </form>
                </Form>
              </div>
            </CardBody>
            <CardBody
              style={
                Object {
                  "borderTop": "1px solid #dfdfdf",
                  "padding": "10px",
                }
              }
            >
              <div
                className="card-body"
                style={
                  Object {
                    "borderTop": "1px solid #dfdfdf",
                    "padding": "10px",
                  }
                }
              >
                <Button
                  active={false}
                  data-test="start-game-btn"
                  disabled={false}
                  type="submit"
                  variant="info"
                >
                  <button
                    className="btn btn-info"
                    data-test="start-game-btn"
                    disabled={false}
                    type="submit"
                  >
                    Start
                  </button>
                </Button>
              </div>
            </CardBody>
          </div>
        </Card>
      </LobbySettings>
    </div>
  </Lobby>
</Provider>
`;
